File: apps/api/internal/models/notification.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Notification is an in-app message surfaced through the admin bell.
// Source distinguishes Sentinel security findings from Pulse perf
// findings from manual operator messages.
type Notification struct {
	ID        string    `gorm:"primarykey;size:36" json:"id"`
	UserID    string    `gorm:"size:36;index" json:"user_id"` // empty = visible to all admins
	Source    string    `gorm:"size:16;index" json:"source"`  // sentinel | pulse | system
	Severity  string    `gorm:"size:16;index" json:"severity"` // critical | high | medium | low | info
	Title     string    `gorm:"size:200" json:"title"`
	Body      string    `gorm:"type:text" json:"body"`
	Link      string    `gorm:"size:500" json:"link"`      // deep-link into /sentinel/ui or /pulse/ui
	Dedup     string    `gorm:"size:128;uniqueIndex" json:"-"` // collision key — repeat firings update Count, not insert
	Count     int       `gorm:"default:1" json:"count"`
	ReadAt    *time.Time `json:"read_at"`
	CreatedAt time.Time `gorm:"index" json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func (n *Notification) BeforeCreate(tx *gorm.DB) error {
	if n.ID == "" {
		n.ID = uuid.New().String()
	}
	return nil
}

File: packages/shared/schemas/notification.ts
import { z } from "zod";

export const CreateNotificationSchema = z.object({
  userId: z.string(),
  source: z.string(),
  severity: z.string(),
  title: z.string(),
  body: z.string(),
  link: z.string(),
  count: z.number().int(),
  readAt: z.string().nullable(),
});

export const UpdateNotificationSchema = z.object({
  userId: z.string().optional(),
  source: z.string().optional(),
  severity: z.string().optional(),
  title: z.string().optional(),
  body: z.string().optional(),
  link: z.string().optional(),
  count: z.number().int().optional(),
  readAt: z.string().nullable(),
});

export type CreateNotificationInput = z.infer<typeof CreateNotificationSchema>;
export type UpdateNotificationInput = z.infer<typeof UpdateNotificationSchema>;
